### PR TITLE
Fixes #31027 - transpiler macros for CoreOS family

### PR DESCRIPTION
**Pull request by lzap**
Adopting from foreman_omaha plugin with some changes. I tried my best to
make those macros and templates as generic and extensible as possible.

Independent follow-up of https://github.com/theforeman/foreman/pull/8042

### CODE REVIEW

File: lib/foreman/renderer/scope/macros/transpilers.rb
Comment (lzap) [inline lib/foreman/renderer/scope/macros/transpilers.rb]: Don't know how to write boolean doc, @ofedoren

File: lib/foreman/command_runner.rb
Comment (lzap) [inline lib/foreman/command_runner.rb:17]: Very strict set of checks, we make sure shell is not spawned as we want to spawn the `ct` process directly to be able to write a good SELinux rules and also to prevent shell-injection.

File: app/views/unattended/provisioning_templates/provision/coreos_provision.erb
Comment (TheKangaroo) [inline app/views/unattended/provisioning_templates/provision/coreos_provision.erb]: I think this will fail for CoreOS/RedHatCoreOS/FedoraCoreOS because the installer has a different name for flatcar.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/CoreOS provision.snap.txt
Comment (TheKangaroo) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/CoreOS provision.snap.txt]: Same problem with the name of the binary here.

File: app/models/setting/provisioning.rb
Comment (ekohl) [inline app/models/setting/provisioning.rb]: I don't think we should use paths in `/var/lib` in our coe. That's IMHO a packaging thing we can solve with a symlink.

Also note that in production `Rails.root` /usr/share/foreman` so it's slightly redundant. `Rails.root` is also a pathname so you can use `Rails.root.join('config', 'ct').to_s` for short.

File: lib/foreman/renderer/configuration.rb
Comment (ekohl) [inline lib/foreman/renderer/configuration.rb]:
```suggestion
        :transpile_fedora_coreos_config,
```

File: lib/foreman/command_runner.rb
Comment (ekohl) [inline lib/foreman/command_runner.rb:13]: Feels like we can ensure this at the settings level

File: app/models/setting/provisioning.rb
Comment (ekohl) [inline app/models/setting/provisioning.rb]: Would it be better to use the name under which development continues? Flatcar?

File: app/views/unattended/provisioning_templates/provision/coreos_provision.erb
Comment (lzap) [reply app/views/unattended/provisioning_templates/provision/coreos_provision.erb]: This template is only meant for legacy CoreOS (thus Flatcar or other forks). But granted this is a leftover. I wonder how I should handle this. Maybe an if-statement, if OS name is Flatcar then use this, otherwise fallback to coreos-install?

File: lib/foreman/command_runner.rb
Comment (lzap) [reply lib/foreman/command_runner.rb:13]: We actually do for this use case but this is an utility method that can be used anywhere in the future.

File: app/models/setting/provisioning.rb
Comment (lzap) [reply app/models/setting/provisioning.rb]: I am open for discussion but it looks like Flatcar is not the only one in the world and CoreOS (or maybe we can decide something like "Legacy CoreOS") is quite a good term.

https://www.coreos-eol.com/

File: lib/foreman/renderer/configuration.rb
Comment (lzap) [reply lib/foreman/renderer/configuration.rb]: lib/foreman/renderer/configuration.rb:47:40: C: Style/TrailingCommaInArrayLiteral: Avoid comma after the last item of an array, unless each item is on its own line.

File: app/models/setting/provisioning.rb
Comment (lzap) [reply app/models/setting/provisioning.rb]: This is the standard practice in GNU but I will remove this to stick with our weird symlinking for easier dev setup.

**Comment by lzap** [comment]
Rebased, the installer command is now correct.

File: lib/foreman/renderer/scope/macros/transpilers.rb
Comment (ofedoren) [reply lib/foreman/renderer/scope/macros/transpilers.rb]: Sorry for the delay. Please, try `optional :validate_input, one_of: [true, false], desc: ...`

Also, you can specify `default: true` or `default: false` after the descriptions.

**Comment by lzap** [comment]
Fixed the snapshot test, I'd appreciate re-reviews and also opinion from @timogoebel / @kamils-iRonin since most of the code is coming from the Omaha plugin (and I'd like users of Ohama to take advantage of these new macros, maybe even templates).

**Comment by timogoebel** [comment]
@laugmanuel: If you have the time do you mind reviewing this? Does this work for our use case?

**Comment by lzap** [comment]
> Does this work for our use case?

In the worst case macros can be used but you can continue using parse-yaml-modify-dump-to-json approach. There are new four macros available now created by @kamils-iRonin - https://github.com/theforeman/foreman/pull/8061

File: app/models/setting/provisioning.rb
Comment (laugmanuel) [inline app/models/setting/provisioning.rb]:
```suggestion
      set('ct_command', N_("CoreOS Transpiler Command"), [which('ct'), '--pretty', '--files-dir', Rails.root.join('config', 'ct').to_s], N_("Full path to CoreOS transpiler (ct) with arguments as an comma-separated array")),
```

File: app/models/setting/provisioning.rb
Comment (laugmanuel) [inline app/models/setting/provisioning.rb]:
```suggestion
      set('fcct_command', N_("Fedora CoreOS Transpiler Command"), [which('fcct'), '--pretty', '--files-dir', Rails.root.join('config', 'ct').to_s], N_("Full path to Fedora CoreOS transpiler (fcct) with arguments as an comma-separated array")),
```

**Review by laugmanuel — COMMENTED** [review]
@timogoebel I can't really assess if this helps / works for us, as I haven't really looked under the hood on how this exactly works.

**Comment by lzap** [comment]
Rebased the typos, thanks. Tests were green, looking good.

**Comment by lzap** [comment]
All is green, @ezr-ondrej

**Comment by nerdalertdk** [comment]
Hi is there an ETA on this pull ?

**Comment by ezr-ondrej** [comment]
I'm sorry for not kiiping the attention to this, I've rebased and I'll merge right after branching if all goes well

**Review by ezr-ondrej — DISMISSED** [review]

**Review by ezr-ondrej — APPROVED** [review]

**Comment by ezr-ondrej** [comment]
Thanks @lzap, @timogoebel, and @ekohl for reviews, thanks @TheKangaroo and @nerdalertdk for testing.
Sorry for keeping you waiting.

**Comment by lzap** [comment]
Thanks all! Would be awesome if you guys could show this feature live on our next demo!

File: app/services/foreman/renderer/scope/macros/transpilers.rb
Comment (ofedoren) [inline app/services/foreman/renderer/scope/macros/transpilers.rb:11]: Forgot to update? :)

**Pull request merged**
